File: 6-plus-one/solution.py
from collections import deque


class Solution:
  def plusOne(self, digits: list[int]) -> list[int]:
    carry = False
    i = len(digits) - 1
    result: deque[int] = deque()
    while carry or i >= 0:
      digit = digits[i] if i >= 0 else 0
      if carry or i == len(digits) - 1:
        digit += 1

      carry = digit >= 10
      result.appendleft(digit if digit < 10 else 0)
      i -= 1

    return list(result)

File: 6-plus-one/solution.cpp
#include <vector>
#include <deque>

class Solution
{
public:
  std::vector<int> plusOne(std::vector<int> &digits)
  {
    bool carry = false;
    int i = digits.size() - 1;
    std::deque<int> result;
    while (carry || i >= 0)
    {
      int digit = i >= 0 ? digits[i] : 0;
      if (carry || i == digits.size() - 1)
      {
        ++digit;
      }

      carry = digit >= 10;
      result.emplace_front(digit < 10 ? digit : 0);
      --i;
    }

    std::vector as_vec(result.begin(), result.end());
    return as_vec;
  }
};